### ISSUE TITLE
wloss_hq_wav2lip_train.py

### ISSUE THREAD

**Issue by NikitaKononov** [primepake/wav2lip_288x288#23]
Hello, I have a question
line 267 in mentioned file
interpolates = alpha * gt + ((1 - alpha) * fake_img)

What is fake_img? It's not defined anywhere, from what variable / func I should get it?

**Comment by primepake** [comment]
I just missed 1 line from my code, I just copied it. Thank you

**Comment by NikitaKononov** [comment]
> I just missed 1 line from my code, I just copied it. Thank you

I've just guessed it 10 minutes ago) gt is ground truth, so g must be fake I thought
thanks)

**Comment by NikitaKononov** [comment]
> I just missed 1 line from my code, I just copied it. Thank you

I think you've missed some imports too, autograd for example

**Comment by primepake** [comment]
ok, imported

**Comment by primepake** [comment]
I also change alpha to constant alpha

**Comment by NikitaKononov** [comment]
> I also change alpha to constant alpha

wasserstein_loss(y_pred, y_true) does not seem to be used anywhere
I should use it instead of perceptual_loss = -pred.mean()?

Without gradient penalty and wass loss (last commit) my tesla v100 was able to handle batch size 16, but now only bs 8 works
Maybe something is not deleted from VRAM?

**Comment by NikitaKononov** [comment]
> I also change alpha to constant alpha

wloss_hq_wav2lip_train.py seems to be constructed from different parts of your work, but it's not obvious to make them work together
for example, I can't understand how gradient penalty affects the training process
and wasserstein loss is not used in the code
cosine loss must be changed too I think, because loss sometimes is negative because of using PReLu

**Comment by primepake** [comment]
I implement based on the original paper, in your case, I need to modify more

**Comment by primepake** [comment]
you should use euclid distance instead of cosine, maybe solve your problem, I've encountered it

**Comment by primepake** [comment]
to understand more about wasserstein loss, read this code: https://github.com/EmilienDupont/wgan-gp/blob/master/training.py

**Comment by primepake** [comment]
I missed mofidied model block, I will public soon

**Comment by NikitaKononov** [comment]
> to understand more about wasserstein loss, read this code: https://github.com/EmilienDupont/wgan-gp/blob/master/training.py

Thanks a lot for your advice, I will try using euclid distance
Oh, this code makes it much clearer. If I've understood right, in our case I should split training for critical model (it's wav2lip) and for generator (it's visual quality discriminator). But where I should place syncnet penalting? Only use syncnet in resulting loss with corresponding weight?

**Comment by primepake** [comment]
don't penalty your syncnet, you need to focus on quality of image and then focus on lip-sync. it will easier to cover 2 tasks at the same time

**Comment by NikitaKononov** [comment]
> don't penalty your syncnet, you need to focus on quality of image and then focus on lip-sync. it will easier to cover 2 tasks at the same time

You've misunderstood me :) I mean that we use SyncNet as kind of penalty for wav2lip, we guide wav2lip to produce lip synced results
So where should I place that?
I've encountered a problem with SyncNet while training previous Wav2liphq. If SyncNet was included in training process too late I wasn't able to get sync loss lower than 2.0
If it was included too early - sync loss wass good, but percep loss didn't go down lower than 0.72
Maybe you have some recipe for best moment to include Syncnet in training process? 
And some advice about disc weight and syncnet weight. By default it's 0.07 and 0.03
Changing those weights affects training, but I can't find a regularity. Maybe you've encountered playing with those weights

**Comment by NikitaKononov** [comment]
> don't penalty your syncnet, you need to focus on quality of image and then focus on lip-sync. it will easier to cover 2 tasks at the same time

Hello, sorry for disturbing you again

I am implementing wgan-gp like training for wav2lip-hq
All goes ok, but I have a question
We find generator loss like this:
d_generated = self.D(generated_data)
g_loss = - d_generated.mean()

How to append syncnet loss correctly?
sync_loss = syncnet(mel, generated_data)
g_loss = -d_generated.mean() * 0.97 + sync_loss * 0.03


is that correct? I am not sure

**Comment by primepake** [comment]
it's up to your distribution of your dataset. as you known critic (discriminator) tried to maximize distance between real and fake and generator is the opposite

**Comment by NikitaKononov** [comment]
> it's up to your distribution of your dataset. as you known critic (discriminator) tried to maximize distance between real and fake and generator is the opposite

with gradient penalty it takes up very large amount of VRAM
I think something excess is stored in the calculation graph, but I don't know exactly what it is
and what I can safely drop from memory
Maybe you have some tips how to minimize VRAM consumption?

**Comment by primepake** [comment]
delete tensor every epoch

**Comment by NikitaKononov** [comment]
> delete tensor every epoch

epoch? maybe you mean each iteration?

**Comment by NikitaKononov** [comment]
> delete tensor every epoch

here's my wgan-gp wav2lip graphs... something goes wrong, But I don't have an idea where should I dig, maybe you've encountered smth like that?

![image](https://user-images.githubusercontent.com/69078939/180929198-ea340394-81d7-4a3a-8074-b68d22be7a76.png)
![image](https://user-images.githubusercontent.com/69078939/180929232-f1b8d170-bcfa-4773-b4c5-08e6df14e34d.png)

**Comment by NikitaKononov** [comment]
> delete tensor every epoch

My wgan gp for wav2lip completely fails to train, searching for the reason
By the way, did your implementation of wav2lip GAN with WLoss and GP performed better, than the original implementation?

if yes, how much quality it added to reslut? is that implementation reasonable?

**Comment by primepake** [comment]
I cannot share my work at here because my commitment, anyway wgan-gp can improve your result. It just improved your image quality. You need to find a way to schedule weight of syncnet, I cann't share all my work. Thanks

**Comment by NikitaKononov** [comment]
> I cannot share my work at here because my commitment, anyway wgan-gp can improve your result. It just improved your image quality. You need to find a way to schedule weight of syncnet, I cann't share all my work. Thanks

I understand you
But I didn't ask you to share your full work or codes
My training fails not because of syncnet, I don't know where to dig
Can you tell what learning rate of D and G did you use, if it's not a commercial secret? And what critical interval
Maybe my problem is there

**Comment by aishoot** [comment]
> > delete tensor every epoch
> 
> My wgan gp for wav2lip completely fails to train, searching for the reason By the way, did your implementation of wav2lip GAN with WLoss and GP performed better, than the original implementation?
> 
> if yes, how much quality it added to reslut? is that implementation reasonable?

I trained the model failed too using wloss_hq_wav2lip_train.py !

**Comment by primepake** [comment]
follow my processing dataset
